cirq-google/cirq_google/devices/grid_device.py

```diff
@@ -120,41 +187,64 @@ def _validate_device_specification(proto: v2.device_pb2.DeviceSpecification) ->
             raise ValueError("Invalid DeviceSpecification: target_ordering cannot be ASYMMETRIC.")
 
 
-def _build_gateset_and_gate_durations(
+def _serialize_gateset_and_gate_durations(
+    out: v2.device_pb2.DeviceSpecification,
+    gateset: cirq.Gateset,
+    gate_durations: Mapping['cirq.GateFamily', 'cirq.Duration'],
+) -> v2.device_pb2.DeviceSpecification:
+    """Serializes the given gateset and gate durations to DeviceSpecification."""
+
+    gate_specs = {}
+    for gate_family in gateset.gates:
+        gate_spec = v2.device_pb2.GateSpecification()
+        gate_rep = next(
+            (gr for gr in _GATES for gf in gr.all_forms if _in_or_equals(gate_family, gf)), None
+        )
+        if gate_rep is None:
+            raise ValueError(f'Unrecognized gate: {gate_family}.')
+
+        # Set gate
+        getattr(gate_spec, gate_rep.gate_spec_name).SetInParent()
+
+        # Set gate duration
+        is_duration_set = False
+        for gf in gate_rep.all_forms:
+            if gf not in gate_durations:
+                continue
+
+            gate_duration_picos = int(gate_durations[gf].total_picos())
+            if is_duration_set and gate_duration_picos != gate_spec.gate_duration_picos:
+                raise ValueError(
+                    'Multiple gate families in the following list exist in the gate duration dict,'
+                    f' and they are expected to have the same duration value: {gate_rep.all_forms}'
+                )
+            is_duration_set = True
+            gate_spec.gate_duration_picos = gate_duration_picos
```

Comment:
Could use a set comprehension to get all durations, then check the size afterward:
```suggestion
        gate_durations_picos = {
            int(gate_durations[gf].total_picos())
            for gf in gate_rep.all_forms
            if gf in gate_durations
        }
        if len(gate_durations_picos) > 1:
            raise ValueError(
                'Multiple gate families in the following list exist in the gate duration dict,'
                f' and they are expected to have the same duration value: {gate_rep.all_forms}'
            )
        elif len(gate_durations_picos) == 1:
            gate_spec.gate_duration_picos = gate_durations_picos.pop()
```